### ISSUE TITLE
allow lists of strings to be stored as fixed length

### ISSUE THREAD

**Issue by bendichter** [NeurodataWithoutBorders/pynwb#1252]
Currently, pynwb saves all vectors of strings as a variable-length vector object. This object is very flexible, but often is not an optimal solution. Compared to the fixed-length counterpart, the variable-length object often is larger and takes longer to read and write. It is also less interoperable with, for instance, the Zarr backend. I would like to be able to instruct pynwb to write string datasets as fixed-length, ideally through the `H5DataIO` interface.

**Comment by t-b** [comment]
@bendichter What is your desired data type in HDF5 terms?

**Comment by oruebel** [comment]
@t-b this would amount essentially to a char array

@bendichter there are a couple of issues to consider here:
- Looking at the h5py docs, fixed length strings are ASCII only http://docs.h5py.org/en/stable/strings.html#fixed-length-ascii
- Using fixed length may lead to loss of data if one later wants to append elements to the dataset with strings that are longer than the chosen fixed type
- If the length of the stings in the array varies, then using fixed length strings causes the strings to be padded with 0s, leading to storage overhead and modification of string values. Whether this is critical is dependent on the values. 

At first glance, from an implementation perspective this would be I think fairly simple to do, i.e., one could simply pass a dtype through with H5DataIO and force the HDF5 backend to use it. However, I think this is dangerous and likely not the right way, because:
1. This approach would enable users to use data types that are incompatible with the spec, hence, creating invalid files,
2. Can lead to silent loss of data due to data conversion on write,
3. Can lead to fatal errors on write due to incompatibly between the desired type and the actual data type.

Addressing the above issues would I believe require a number of deeper changes. I'm not 100% sure how to do this the right way, but one probably would need to:

1. The dtype should probably be set in construction of the Builders (not in the IO backend), so the build process would need to be modified to (a) look at DataIO and (b) make sure the data types are compatible with the spec and (c) the DataIO base class would need to add a dtype parameter (rather than just H5DataIO
2. The data conversion would also need to be handled (and warning/errors) issued in the build process
3. All this could get even more tricky when dealing with iterative write. 

Two ways around this problem may be:
1) One may be able to avoid some of the trouble by restricting the setting of dtypes to strings only to reduce the number of edge cases one would need to deal with.
1) Rather than trying to fix this post-hoc on write, it may be simpler to fix this in the schema and have a string and char data type

**Comment by bendichter** [comment]
Yes, fixed length strings are a bit less flexible. This is why I'm not pushing for this type by default, but as a user option. That's why I proposed putting it in `HDF5IO`. I don't see how a user would change Builder behavior, so I would guess that changes to the Builder would change default behavior, which is not what I want, for the reasons you mentioned.

> One may be able to avoid some of the trouble by restricting the setting of dtypes to strings only to reduce the number of edge cases one would need to deal with.

I am imagining some added flexibility in the way data is stored for specific data types, but not total flexibility to use any representation you want for any data, so restricting this makes sense to me.

> Rather than trying to fix this post-hoc on write, it may be simpler to fix this in the schema and have a string and char data type

This is fine with me, especially if we have a term that allows either. I would still need a way to control the data type, because I want this in a custom `DynamicTable` column.

**Comment by oruebel** [comment]
> because I want this in a custom `DynamicTable` column.

For this, as a test I would try to set the column to a numpy array with ``dtype='S10'``. I think since this is a custom column that dtype should probably be just passed through as is, i.e., I think this should already do what you want.

**Comment by t-b** [comment]
@oruebel Thanks for the explanation. I  think supporting a new string datatype which is ASCII-only is not the right direction.

**Comment by bendichter** [comment]
@oruebel I tried what you suggested and it's not working for me, though I may not be doing it right:

```python
from pynwb import NWBFile, NWBHDF5IO
import numpy as np
import datetime
from pynwb.core import DynamicTable

#string_data = ['aa', 'bb']  # works
string_data = np.array(['aa', 'bb'], dtype='S10')  # does not work
#string_data = np.array(['aa', 'bb'], dtype=np.unicode)  # does not work


nwbfile = NWBFile(session_description='test foraging2',
                  identifier='behavior_session_uuid',
                  session_start_time=datetime.datetime.now())

mod = nwbfile.create_processing_module('test','test')
dt = DynamicTable('name', 'desc', id=[0,1])
dt.add_column('test_S10','desc', data=string_data)
mod.add(dt)

with NWBHDF5IO('/Users/bendichter/dev/sandbox/test_S10.nwb', 'w') as io:
    io.write(nwbfile)
```

error for `np.array(['aa', 'bb'], dtype='S10')`:
```
~/dev/hdmf/src/hdmf/backends/hdf5/h5tools.py in write_dataset(self, **kwargs)
    956             # Write a regular in memory array (e.g., numpy array, list etc.)
    957             elif hasattr(data, '__len__'):
--> 958                 dset = self.__list_fill__(parent, name, data, options)
    959             # Write a regular scalar dataset
    960             else:

~/dev/hdmf/src/hdmf/backends/hdf5/h5tools.py in __list_fill__(cls, parent, name, data, options)
   1117             msg = "Could not create dataset %s in %s with shape %s, dtype %s, and iosettings %s. %s" % \
   1118                   (name, parent.name, str(data_shape), str(dtype), str(io_settings), str(exc))
-> 1119             raise Exception(msg) from exc
   1120         # Write the data
   1121         if len(data) > dset.shape[0]:

Exception: Could not create dataset test_S10 in /processing/test/name with shape (2,), dtype <class 'numpy.bytes_'>, and iosettings {}. Size must be positive (size must be positive)
```

and then for good measure I tried using the unicode type, which also did not work (but I don't think I should expect it to).

**Comment by oruebel** [comment]
@bendichter it looks like this may be a bug in how string data is handled. It looks like in the build process the data type on the DatasetBuilder is set to ``bytes`` in this  case. I.e., instead of trying to create the dataset with the ``'S10'`` data type, ``create_dataset`` is being called with ``dtype=np.bytes_`` and h5py does not seem to like that data type. I have not had a chance to dig any deeper on this, but this suggest to me that there is a bug in how fixed-length string arrays are being handled on write, i.e., at the very least I think this should not fail but write it as a bytes array. I would suggest to create an issue on the HDMF repo for this.

**Comment by oruebel** [comment]
@bendichter I've created a simple example with just HDMF and as suspected, it looks like this is an issue in HDMF. I've created the following bug report on HDMF https://github.com/hdmf-dev/hdmf/issues/386

**Comment by oruebel** [comment]
https://github.com/hdmf-dev/hdmf/pull/387

**Comment by oruebel** [comment]
@bendichter is this working now that the HDMF PR has been merged?

**Comment by bendichter** [comment]
@oruebel I haven't gotten to testing this yet. We ended up writing the necessary code to read var length strings in Zarr, but I still think this is useful, as it takes up much less space than var-length strings.
